chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,354 @@
+
+	***** END LICENSE BLOCK *****
+*/
+
+const { COLUMNS: ITEMTREE_COLUMNS } = require("zotero/itemTreeColumns");
+
+/**
+ * @typedef {import("../itemTreeColumns.jsx").ItemTreeColumnOptions} ItemTreeColumnOptions
+ * @typedef {"dataKey" | "label" | "pluginID"} RequiredCustomColumnOptionKeys
+ * @typedef {Required<Pick<ItemTreeColumnOptions, RequiredCustomColumnOptionKeys>>} RequiredCustomColumnOptionsPartial
+ * @typedef {Omit<ItemTreeColumnOptions, RequiredCustomColumnOptionKeys>} CustomColumnOptionsPartial
+ * @typedef {RequiredCustomColumnOptionsPartial & CustomColumnOptionsPartial} ItemTreeCustomColumnOptions
+ * @typedef {Partial<Omit<ItemTreeCustomColumnOptions, "enabledTreeIDs">>} ItemTreeCustomColumnFilters
+ */
+
+class ItemTreeManager {
+	_observerAdded = false;
+
+	/** @type {Record<string, ItemTreeCustomColumnOptions}} */
+	_customColumns = {};
+
+	/**
+	 * Register a custom column. All registered columns must be valid, and must have a unique dataKey.
+	 * Although it's async, resolving does not promise the item trees are updated.
+	 * @param {ItemTreeCustomColumnOptions | ItemTreeCustomColumnOptions[]} options - An option or array of options to register
+	 * @returns {string | string[] | false} - The dataKey(s) of the added column(s) or false if no columns were added
+	 * @example
+	 * A minimal custom column:
+	 * ```js
+	 * // You can unregister the column later with Zotero.ItemTreeManager.unregisterColumns(registeredDataKey);
+	 * const registeredDataKey = await Zotero.ItemTreeManager.registerColumns(
+	 * {
+	 *     dataKey: 'rtitle',
+	 *     label: 'reversed title',
+	 *     pluginID: 'make-it-red@zotero.org', // Replace with your plugin ID
+	 *     dataProvider: (item, dataKey) => {
+	 *         return item.getField('title').split('').reverse().join('');
+	 *     },
+	 * });
+	 * ```
+	 * @example
+	 * A custom column using all available options.
+	 * Note that the column will only be shown in the main item tree.
+	 * ```js
+	 * const registeredDataKey = await Zotero.ItemTreeManager.registerColumns(
+	 * {
+	 *     dataKey: 'rtitle',
+	 *     label: 'reversed title',
+	 *     enabledTreeIDs: ['main'], // only show in the main item tree
+	 *     defaultSort: 1, // sort by increasing order
+	 *     flex: 0, // don't take up all available space
+	 *     width: 100, // assign fixed width in pixels
+	 *     fixedWidth: true, // don't allow user to resize
+	 *     staticWidth: true, // don't allow coloumn to be resized when the tree is resized
+	 *     minWidth: 50, // minimum width in pixels
+	 *     // iconPath: 'chrome://zotero/skin/tick.png', // icon to show in the column header
+	 *     htmlLabel: '<span style="color: red;">reversed title</span>', // use HTML in the label. This will override the label and iconPath property
+	 *     ignoreInColumnPicker: false, // show in the column picker
+	 *     submenu: true, // show in the column picker submenu
+	 *     primary: false, // only one primary column is allowed
```

Comment:
Would a plugin ever use `primary: true`?

Reply:
we don't want them to, I guess.

Reply:
So should we remove this from here? Or leave it undocumented? I forget where we landed in terms of whether we'd use this API ourselves.

Reply:
We wouldn't use this API ourselves, so leave undocumented and deny attempts to register.

Reply:
Added explaination to let people know they should not set this to true